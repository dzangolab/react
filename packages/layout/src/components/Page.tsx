import React from "react";

interface Properties {
  subTitle?: string;
  title?: string;
  toolbar?: React.ReactNode;
}

const Page = ({ subTitle, title, toolbar }: Properties) => {
  return (
    <div className="page">
      {title && (
        <h1>
          {title}
          {subTitle && <small>{subTitle}</small>}
        </h1>
      )}
      {toolbar && <div className="toolbar">{toolbar}</div>}
    </div>
  );
};

export default Page;
