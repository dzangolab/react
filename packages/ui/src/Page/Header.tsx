interface IHeaderProperties {
  title?: string;
  subtitle?: React.ReactNode | string;
  toolbar?: React.ReactNode;
  breadcrumb?: React.ReactNode;
}

export const PageHeader = ({
  title,
  subtitle,
  toolbar,
  breadcrumb,
}: IHeaderProperties) => {
  return (
    <div className="page-header">
      {breadcrumb ? <div className="breadcrumb">{breadcrumb}</div> : null}
      {title && (
        <h1>
          {title}
          {typeof subtitle === "string" ? <small>{subtitle}</small> : subtitle}
        </h1>
      )}
      {toolbar && <div className="toolbar">{toolbar}</div>}
    </div>
  );
};
