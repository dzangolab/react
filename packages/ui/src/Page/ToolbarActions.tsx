import React, { useState } from "react";

import { Button, DropdownMenu, IButtonProperties, MenuItem } from "..";
import { ConfirmationModal, IModalProperties } from "../ConfirmationModal";

export interface ActionsMenuItem
  extends Omit<MenuItem, "onClick">,
    Partial<IButtonProperties> {
  requireConfirmationModal?: boolean;
  onClick?: (event?: React.MouseEvent | React.KeyboardEvent) => void;
}

export interface ToolbarActionsMenuProperties {
  actions?: ActionsMenuItem[];
}

export const ToolbarActions = ({ actions }: ToolbarActionsMenuProperties) => {
  const [confirmation, setConfirmation] = useState<IModalProperties | null>();

  const renderActions = () => {
    if (!actions || !actions.length) {
      return null;
    }

    const { icon, label, onClick, disabled, ...rest } = actions[0];

    if (actions.length == 1 && icon) {
      return (
        <Button
          variant="textOnly"
          iconLeft={icon}
          data-pr-tooltip={label}
          onClick={onClick}
          rounded
          {...rest}
        />
      );
    }

    return (
      <DropdownMenu
        label={<Button iconLeft="pi pi-cog" size="medium" variant="outlined" />}
        menu={actions}
        hideDropdownIcon
      />
    );
  };

  return (
    <>
      {renderActions()}
      {!!confirmation && (
        <ConfirmationModal {...confirmation} visible={!!confirmation} />
      )}
    </>
  );
};
