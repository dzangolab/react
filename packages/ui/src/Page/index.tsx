import React from "react";

import LoadingPage, { LoadingPageProperties } from "../LoadingPage";

interface Properties
  extends Pick<
    React.DetailedHTMLProps<
      React.HTMLAttributes<HTMLDivElement>,
      HTMLDivElement
    >,
    "aria-orientation"
  > {
  children?: React.ReactNode;
  className?: string;
  errorMessage?: string;
  loading?: boolean;
  loadingComponent?: React.ReactElement;
  loadingPageStyle?: LoadingPageProperties;
  subtitle?: React.ReactNode;
  title?: string;
  toolbar?: React.ReactNode;
}

const Page: React.FC<Properties> = ({
  children,
  className,
  errorMessage,
  loading = false,
  loadingComponent,
  loadingPageStyle,
  subtitle,
  title,
  toolbar,
  ...others
}: Properties) => {
  let child = null;
  let _className = "page";
  console.log(typeof subtitle);
  if (loading) {
    child = loadingComponent ? (
      loadingComponent
    ) : (
      <LoadingPage {...loadingPageStyle} />
    );
  } else if (errorMessage) {
    child = (
      <div className="error" role="alert">
        <span>{errorMessage}</span>
      </div>
    );
  }

  if (className) {
    _className += ` ${className}`;
  }

  return (
    <div className={_className}>
      {title && (
        <h1>
          {title}
          {typeof subtitle === "string" ? (
            <small>{subtitle}</small>
          ) : (
            <span>{subtitle}</span>
          )}
        </h1>
      )}
      {toolbar && <div className="toolbar">{toolbar}</div>}
      <div data-testid="page-content" className="content" {...others}>
        {child ? child : children}
      </div>
    </div>
  );
};

export default Page;
