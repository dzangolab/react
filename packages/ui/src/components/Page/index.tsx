import React from "react";

import { LoadingPage } from "../Loading";

import "./page.css";

interface Properties {
  children?: React.ReactNode;
  errorMessage?: string;
  loading?: boolean;
  loadingComponent?: React.ReactElement;
  subtitle?: string;
  title?: string;
  toolbar?: React.ReactNode;
}

const Page: React.FC<Properties> = ({
  children,
  errorMessage,
  loading = false,
  loadingComponent = <LoadingPage />,
  subtitle,
  title,
  toolbar,
}: Properties) => {
  let child = null;

  if (loading) {
    child = loadingComponent;
  } else if (errorMessage) {
    child = (
      <div className="page">
        <div role="alert">
          <span>{errorMessage}</span>
        </div>
      </div>
    );
  } else {
    child = (
      <div className="page">
        {title && (
          <h1>
            {title}
            {subtitle && <small>{subtitle}</small>}
          </h1>
        )}
        {toolbar && <div className="toolbar">{toolbar}</div>}
        <div data-testid="page-content" className="content">
          {children}
        </div>
      </div>
    );
  }
  return child;
};

export default Page;
