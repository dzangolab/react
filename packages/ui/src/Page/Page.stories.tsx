import { Meta, StoryFn } from "@storybook/react";
import React from "react";

import Page from "./index";

type PageType = typeof Page;
type PageProperties = Meta<PageType>;

const PageStory: PageProperties = {
  title: "UI/Page",
  component: Page,
};

const Template: StoryFn<PageType> = (arguments_) => <Page {...arguments_} />;

export const Loading: PageProperties = Template.bind({});
export const CustomLoadingComponent: PageProperties = Template.bind({});
export const Error: PageProperties = Template.bind({});
export const TitleAndSubtitle: PageProperties = Template.bind({});
export const Toolbar: PageProperties = Template.bind({});

Loading.args = {
  loading: true,
};

CustomLoadingComponent.args = {
  loading: true,
  loadingComponent: <span>Loading...</span>,
};

TitleAndSubtitle.args = {
  title: "This is the title of this page",
  subtitle: "This is the subtitle of this page",
};

Toolbar.args = {
  toolbar: (
    <>
      <button>Action one</button>
      <button>Action two</button>
    </>
  ),
};

Error.args = {
  errorMessage: "404: Not found",
};

export default PageStory;
