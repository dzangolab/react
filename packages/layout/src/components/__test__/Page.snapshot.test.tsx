import renderer from "react-test-renderer";
import { expect, test } from "vitest";

import Page from "../Page";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = renderer.create(
    <Page
      title="Test title"
      subTitle="Test subtitle"
      toolbar={<div>Toolbar content</div>}
    />
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
