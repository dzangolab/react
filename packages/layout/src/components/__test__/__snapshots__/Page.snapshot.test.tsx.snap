// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<div
  className="page"
>
  <h1>
    Test title
    <small>
      Test subtitle
    </small>
  </h1>
  <div
    className="toolbar"
  >
    <div>
      Toolbar content
    </div>
  </div>
</div>
`;
